drivers/i2c/i2c_esp32.c

```diff
@@ -0,0 +1,781 @@
+					 bool setting)
+{
+	volatile u32_t *periph_clk_reg = (u32_t *)DPORT_PERIP_CLK_EN_REG;
+	volatile u32_t *periph_rst_reg = (u32_t *)DPORT_PERIP_RST_EN_REG;
+
+	if (setting) {
+		*periph_clk_reg |= config->enable_bits.clk;
+		*periph_rst_reg &= ~config->enable_bits.rst;
+	} else {
+		*periph_clk_reg &= ~config->enable_bits.clk;
+		*periph_rst_reg |= config->enable_bits.rst;
+	}
+}
+
+static const char *i2c_esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+
+static int i2c_esp32_configure_pins(int pin, bool pull_up, bool default_level,
+				    int matrix_out, int matrix_in)
+{
+	const char *device_name = i2c_esp32_get_gpio_for_pin(pin);
+	struct device *gpio;
+
+	if (!device_name) {
+		return -EINVAL;
+	}
+	gpio = device_get_binding(device_name);
+	if (!gpio) {
+		return -EINVAL;
+	}
+
+	gpio_pin_configure(gpio, pin,
+			   GPIO_DIR_OUT | GPIO_DS_DISCONNECT_LOW |
+			   pull_up ? GPIO_PUD_PULL_UP : GPIO_PUD_NORMAL);
+	gpio_pin_write(gpio, pin, default_level);
+
+	gpio_matrix_out(pin, matrix_out, 0, 0);
+	gpio_matrix_in(pin, matrix_in, 0);
+
+	return 0;
+}
+
+static void i2c_esp32_configure_frequency(const struct i2c_esp32_config *config,
+					  u32_t freq_hz)
+{
+	const u32_t period = (APB_CLK_FREQ / freq_hz) / 2;
+
+	*config->regs.scl_low_period = period << I2C_SCL_LOW_PERIOD_S;
+	*config->regs.scl_high_period = period << I2C_SCL_HIGH_PERIOD_S;
+
+	*config->regs.scl_start_h_time =
+		(period / 2) << I2C_SCL_START_HOLD_TIME_S;
+	*config->regs.scl_rstart_s_time =
+		(period / 2) << I2C_SCL_RSTART_SETUP_TIME_S;
+	*config->regs.scl_stop_h_time =
+		(period / 2) << I2C_SCL_STOP_HOLD_TIME_S;
+	*config->regs.scl_stop_s_time =
+		(period / 2) << I2C_SCL_STOP_SETUP_TIME_S;
+
+	*config->regs.sda_hold_time = (period / 4) << I2C_SDA_HOLD_TIME_S;
+	*config->regs.sda_sample_time = (period / 4) << I2C_SDA_SAMPLE_TIME_S;
+}
+
+static int i2c_esp32_configure(struct device *dev, u32_t dev_config_raw)
+{
+	const struct i2c_esp32_config *config = dev->config->config_info;
+	struct i2c_esp32_data *data = dev->driver_data;
+	union dev_config dev_config = { .raw = dev_config_raw };
+	int key = irq_lock();
+	u32_t v;
+	int r;
+
+	r = i2c_esp32_configure_pins(config->pins.scl,
+				     config->pins.scl_pull_up,
+				     config->pins.default_level,
+				     config->sig.scl_out,
+				     config->sig.scl_in);
+	if (r < 0) {
+		return r;
+	}
+
+	r = i2c_esp32_configure_pins(config->pins.sda,
+				     config->pins.sda_pull_up,
+				     config->pins.default_level,
+				     config->sig.sda_out,
+				     config->sig.sda_in);
+	if (r < 0) {
+		return r;
+	}
+
+	i2c_esp32_set_device_enabled(config, true);
+
+	v = *config->regs.ctr;
+
+	/* MSB or LSB first is configurable for both TX and RX */
```

Comment:
So... do we care?  Gotta be honest I didn't know configurable bit order was even a thing for I2C.  Are there any peripherals that actually care or want this?  Maybe drop this feature, or at least make it a sanely defaulted kconfig that most users don't ever touch.

Reply:
I've abused these things in the past.  Once I wanted to control an array of WS2812B LEDs using Linux, but it's very difficult to obtain precise timings, specially when you have hundreds of daisy-chained LEDs.  By using the [SPI controller and framing the bits just right](https://hackaday.com/2014/09/10/driving-ws2812b-pixels-with-dma-based-spi/), it's possible to convince it can speak the WS2812B protocol.  And sometimes, doing things like this requires doing things that are not standard.

Not, to answer your question: I don't have a clue if we care.